alf/algorithms/goal_generator.py

```diff
@@ -0,0 +1,79 @@
+# Copyright (c) 2020 Horizon Robotics. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import gin.tf
+import tensorflow as tf
+import numpy as np
+import functools
+
+import tf_agents.specs.tensor_spec as tensor_spec
+from tf_agents.trajectories.time_step import StepType
+
+from alf.algorithms.algorithm import Algorithm, AlgorithmStep
+from alf.data_structures import ActionTimeStep, PolicyStep, TrainingInfo, LossInfo, namedtuple
+
+import alf.utils.common as common
+
+GoalInfo = namedtuple("GoalInfo", ["goal"], default_value=())
+
+
+@gin.configurable
+class RandomCategoricalGoalGenerator(Algorithm):
+    """Random Goal Generation Module
+
+    This module generates a random categorical goal for the agent.
+    """
+
+    def __init__(self,
+                 num_of_goals,
+                 goal_feature_size,
```

Comment:
I suggest that here we only have one arg goal_spec, and it can be tf.TensorSpec((), tf.int32) or tf.TensorSpec([size], tf.float32) or other form. You can set this spec in a gin file